OrbitCaptureClient/CaptureEventProcessorTest.cpp

```diff
@@ -0,0 +1,614 @@
+
+  CaptureEvent event;
+  SchedulingSlice* scheduling_slice = event.mutable_scheduling_slice();
+  scheduling_slice->set_core(2);
+  scheduling_slice->set_pid(42);
+  scheduling_slice->set_tid(24);
+  scheduling_slice->set_in_timestamp_ns(3);
+  scheduling_slice->set_out_timestamp_ns(100);
+
+  TimerInfo actual_timer;
+  EXPECT_CALL(listener, OnTimer).Times(1).WillOnce(SaveArg<0>(&actual_timer));
+
+  event_processor.ProcessEvent(event);
+
+  EXPECT_EQ(actual_timer.start(), scheduling_slice->in_timestamp_ns());
+  EXPECT_EQ(actual_timer.end(), scheduling_slice->out_timestamp_ns());
+  EXPECT_EQ(actual_timer.process_id(), scheduling_slice->pid());
+  EXPECT_EQ(actual_timer.thread_id(), scheduling_slice->tid());
+  EXPECT_EQ(actual_timer.processor(), scheduling_slice->core());
+  EXPECT_EQ(actual_timer.type(), TimerInfo::kCoreActivity);
+}
+
+TEST(CaptureEventProcessor, CanHandleOneCallstackSample) {
+  MockCaptureListener listener;
+  CaptureEventProcessor event_processor(&listener);
+
+  CaptureEvent event;
+  CallstackSample* callstack_sample = event.mutable_callstack_sample();
+  callstack_sample->set_pid(1);
+  callstack_sample->set_tid(3);
+  Callstack* callstack = callstack_sample->mutable_callstack();
+  callstack->add_pcs(14);
+  callstack->add_pcs(15);
+  callstack_sample->set_timestamp_ns(100);
+
+  CallStack actual_call_stack;
+  EXPECT_CALL(listener, OnUniqueCallStack).Times(1).WillOnce(SaveArg<0>(&actual_call_stack));
+  CallstackEvent actual_call_stack_event;
+  EXPECT_CALL(listener, OnCallstackEvent).Times(1).WillOnce(SaveArg<0>(&actual_call_stack_event));
+
+  event_processor.ProcessEvent(event);
+
+  EXPECT_EQ(actual_call_stack_event.time(), callstack_sample->timestamp_ns());
+  EXPECT_EQ(actual_call_stack_event.thread_id(), callstack_sample->tid());
+  EXPECT_EQ(actual_call_stack_event.callstack_hash(), actual_call_stack.GetHash());
+  ASSERT_EQ(actual_call_stack.GetFramesCount(), callstack->pcs_size());
+  for (size_t i = 0; i < actual_call_stack.GetFramesCount(); ++i) {
+    EXPECT_EQ(actual_call_stack.GetFrame(i), callstack->pcs(i));
+  }
+}
+
+TEST(CaptureEventProcessor, WillOnlyHandleUniqueCallstacksOnce) {
+  MockCaptureListener listener;
+  CaptureEventProcessor event_processor(&listener);
+  std::vector<CaptureEvent> events;
+
+  CaptureEvent event_1;
+  CallstackSample* callstack_sample_1 = event_1.mutable_callstack_sample();
+  callstack_sample_1->set_pid(1);
+  callstack_sample_1->set_tid(3);
+  Callstack* callstack_1 = callstack_sample_1->mutable_callstack();
+  callstack_1->add_pcs(14);
+  callstack_1->add_pcs(15);
+  callstack_sample_1->set_timestamp_ns(100);
+
+  CaptureEvent event_2;
+  CallstackSample* callstack_sample_2 = event_2.mutable_callstack_sample();
+  callstack_sample_2->set_pid(2);
+  callstack_sample_2->set_tid(4);
+  Callstack* callstack_2 = callstack_sample_2->mutable_callstack();
+  callstack_2->add_pcs(14);
+  callstack_2->add_pcs(15);
+  callstack_sample_2->set_timestamp_ns(200);
```

Comment:
You could copy the event_1 and then change the three deviating properties. I think that would make it easier to read and grasp.

Reply:
That should be obsolete due to the init function.